src/main/java/com/riscure/trs/parameter/trace/TraceParameterMap.java

```diff
@@ -33,54 +37,157 @@ public static TraceParameterMap deserialize(byte[] bytes, TraceParameterDefiniti
     }
 
     public void put(String key, byte value) {
-        put(key, new byte[]{value});
+        put(key, TraceParameterFactory.create(Byte.class, value));
     }
 
     public void put(String key, byte[] value) {
-        put(key, new ByteArrayParameter(value));
+        put(key, TraceParameterFactory.create(byte[].class, value));
     }
 
     public void put(String key, short value) {
-        put(key, new short[]{value});
+        put(key, TraceParameterFactory.create(Short.class, value));
     }
 
     public void put(String key, short[] value) {
-        put(key, new ShortArrayParameter(value));
+        put(key, TraceParameterFactory.create(short[].class, value));
     }
 
     public void put(String key, int value) {
-        put(key, new int[]{value});
+        put(key, TraceParameterFactory.create(Integer.class, value));
     }
 
     public void put(String key, int[] value) {
-        put(key, new IntArrayParameter(value));
+        put(key, TraceParameterFactory.create(int[].class, value));
     }
 
     public void put(String key, float value) {
-        put(key, new float[]{value});
+        put(key, TraceParameterFactory.create(Float.class, value));
     }
 
     public void put(String key, float[] value) {
-        put(key, new FloatArrayParameter(value));
+        put(key, TraceParameterFactory.create(float[].class, value));
     }
 
     public void put(String key, long value) {
-        put(key, new long[]{value});
+        put(key, TraceParameterFactory.create(Long.class, value));
     }
 
     public void put(String key, long[] value) {
-        put(key, new LongArrayParameter(value));
+        put(key, TraceParameterFactory.create(long[].class, value));
     }
 
     public void put(String key, double value) {
-        put(key, new double[]{value});
+        put(key, TraceParameterFactory.create(Double.class, value));
     }
 
     public void put(String key, double[] value) {
-        put(key, new DoubleArrayParameter(value));
+        put(key, TraceParameterFactory.create(double[].class, value));
     }
 
     public void put(String key, String value) {
-        put(key, new StringParameter(value));
+        put(key, TraceParameterFactory.create(String.class, value));
+    }
+
+    public <T> void put(TypedKey<T> typedKey, T value) {
+        put(typedKey.getKey(), TraceParameterFactory.create(typedKey.getCls(), value));
+    }
+
+    public <T> T get(TypedKey<T> typedKey) {
+        if (get(typedKey.getKey()).length() == 1) {
+            return typedKey.getCls().cast(get(typedKey.getKey()).getSimpleValue());
+        } else {
+            return typedKey.getCls().cast(get(typedKey.getKey()).getValue());
```

Comment:
You can make this slightly easier to read by extracting a variable:
```suggestion
        TraceParameter traceParameter = get(typedKey.getKey());
        if (traceParameter.length() == 1) {
            return typedKey.getCls().cast(traceParameter.getSimpleValue());
        } else {
            return typedKey.getCls().cast(traceParameter.getValue());
        }
```